game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/air/AirDefendVsNonSubsStepTest.java

```diff
@@ -6,67 +6,59 @@
 import java.util.List;
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 import org.mockito.junit.jupiter.MockitoExtension;
 
 @ExtendWith(MockitoExtension.class)
 class AirDefendVsNonSubsStepTest {
 
-  @Test
-  @DisplayName("valid() is true if defender has air and no destroyer and attacker has sub")
-  void airVsSub() {
-    final Unit defender1 = givenUnit();
-    final Unit defender2 = givenUnitIsAir();
-    final Unit attacker = givenUnitCanNotBeTargetedBy(mock(UnitType.class));
-
-    final BattleState battleState =
-        givenBattleState()
-            .attackingUnits(List.of(defender1, defender2))
-            .defendingUnits(List.of(attacker))
-            .build();
-    final AirAttackVsNonSubsStep underTest = new AirAttackVsNonSubsStep(battleState);
-    assertThat(underTest.valid(), is(true));
+  @ParameterizedTest(name = "[{index}] {0} is {2}")
+  @MethodSource
+  void testWhatIsValid(
+      final String displayName, final BattleState battleState, final boolean expected) {
+    final AirDefendVsNonSubsStep underTest = new AirDefendVsNonSubsStep(battleState);
+    assertThat(underTest.valid(), is(expected));
+    if (expected) {
+      assertThat(underTest.getNames(), hasSize(1));
+    } else {
+      assertThat(underTest.getNames(), hasSize(0));
+    }
   }
 
-  @Test
-  @DisplayName("valid() is false if defender has air and destroyer")
-  void airDestroyerVsAnything() {
-    final Unit defender1 = givenUnitDestroyer();
-    final Unit defender2 = givenUnitIsAir();
-    // once a destroyer is around, it doesn't care about the attacker units
-    final Unit attacker = mock(Unit.class);
-
-    final BattleState battleState =
-        givenBattleState()
-            .attackingUnits(List.of(defender1, defender2))
-            .defendingUnits(List.of(attacker))
-            .build();
-    final AirAttackVsNonSubsStep underTest = new AirAttackVsNonSubsStep(battleState);
-    assertThat(underTest.valid(), is(false));
-  }
-
-  @Test
-  @DisplayName("valid() is false if defender has air and no destroyer and attacker has no sub")
-  void airVsOther() {
-    final Unit defender1 = givenUnit();
-    final Unit defender2 = givenUnitIsAir();
-    final Unit attacker = givenUnit();
-
-    final BattleState battleState =
-        givenBattleState()
-            .attackingUnits(List.of(defender1, defender2))
-            .defendingUnits(List.of(attacker))
-            .build();
-    final AirAttackVsNonSubsStep underTest = new AirAttackVsNonSubsStep(battleState);
-    assertThat(underTest.valid(), is(false));
+  static List<Arguments> testWhatIsValid() {
+    return List.of(
+        Arguments.of(
+            "Defender has air units and no destroyers vs Attacker subs",
+            givenBattleState()
```

Comment:
The givenBattleState is still odd, it probably should at least return a builder().

You can make list values non-file and use @Builder.Default to initialize them to empty lists.
Also consider using `@Singular` which I think will default the value to an empty list if not specified. Using `@Singular` on the attacking and defending units properties, you could instead write:
```
MockBattleState.builder()
   .attackingUnit(givenUnitCanNotBeTargetedBy(mock(UnitType.class)))
   .defendingUnit(givenUnit())
   .defendingUnit(givenUnitIsAir())
   .build();
```

Reply:
Thanks for the pointer to `@Singular`

Reply:
`@Singular` is causing the `Error Prone static analysis plugin` to throw an exception.  All you need to do is put `@Singular` on one of the fields in `MockBattleState` and then run it.

Reply:
Ah yeah, error prone and singular do not play well.  I think the error prone analysis that fails needs to be suppressed for classes that use singular. Perhaps turning the rule off is a better move. IMO `@Singular` can be really useful, error prone is arguably getting in our way with it.